java-checks/src/main/java/org/sonar/java/checks/spring/SpringScanDefaultPackageCheck.java

```diff
@@ -0,0 +1,171 @@
+    Set<String> scanPackageAttributeNames = SCAN_PACKAGE_ATTRIBUTES.get(annotation.symbolType().fullyQualifiedName());
+    if (scanPackageAttributeNames != null) {
+      List<ExpressionTree> scanPackageAttributes = annotation.arguments().stream()
+        .filter(argument -> scanPackageAttributeNames.contains(attributeName(argument)))
+        .collect(Collectors.toList());
+
+      if (scanPackageAttributes.isEmpty()) {
+        checkIfAnnotationIsInDefaultPackage(annotation);
+      } else {
+        checkAnnotationPackageAttributes(annotation, scanPackageAttributes);
+      }
+    }
+  }
+
+  private void checkIfAnnotationIsInDefaultPackage(AnnotationTree annotation) {
+    if (isNodeInDefaultPackage(annotation)) {
+      reportIssue(annotation.annotationType(), "Remove the annotation \"@" + annotation.symbolType().name() +
+        "\" or move the annotated class out of the default package.");
+    }
+  }
+
+  private void checkAnnotationPackageAttributes(AnnotationTree annotation, List<ExpressionTree> scanPackageAttributes) {
+    List<ExpressionTree> values = scanPackageAttributes.stream()
+      .flatMap(SpringScanDefaultPackageCheck::extractValues)
+      .collect(Collectors.toList());
+
+    values.stream()
+      .flatMap(SpringScanDefaultPackageCheck::findEmptyString)
+      .forEach(expression -> reportIssue(expression, "Define packages to scan. Don't rely on the default package."));
+
+    values.stream()
+      .flatMap(SpringScanDefaultPackageCheck::findClassInDefaultPackage)
+      .forEach(identifier -> reportIssue(identifier, "Remove the annotation \"@" + annotation.symbolType().name() +
+        "\" or move the \"" + identifier.name() + "\" class out of the default package."));
+  }
+
+  private static String attributeName(ExpressionTree expression) {
+    if (expression.is(Tree.Kind.ASSIGNMENT)) {
+      AssignmentExpressionTree assignment = (AssignmentExpressionTree) expression;
+      if (assignment.variable().is(Tree.Kind.IDENTIFIER)) {
+        return ((IdentifierTree) assignment.variable()).name();
+      }
+    }
+    return DEFAULT_ATTRIBUTE;
+  }
+
+  private static Stream<ExpressionTree> extractValues(ExpressionTree argument) {
+    ExpressionTree expression = argument;
+    if (expression.is(Tree.Kind.ASSIGNMENT)) {
+      expression = ((AssignmentExpressionTree) expression).expression();
+    }
+    if (expression.is(Tree.Kind.NEW_ARRAY)) {
+      return ((NewArrayTree) expression).initializers().stream()
+        .flatMap(SpringScanDefaultPackageCheck::extractValues);
+    }
+    return Stream.of(expression);
+  }
+
+  private static Stream<ExpressionTree> findEmptyString(ExpressionTree expression) {
+    String stringValue = ConstantUtils.resolveAsStringConstant(expression);
+    if (stringValue != null && stringValue.isEmpty()) {
+      return Stream.of(expression);
+    }
+    return Stream.empty();
+  }
+
+  private static Stream<IdentifierTree> findClassInDefaultPackage(ExpressionTree expression) {
```

Comment:
Moreover another issue here is that we could imagine to have mutiple issues on the same annotation and that does not seem to be covered in testcases. (and is it something expected).

Reply:
Now it is covered.